### PR TITLE
LogProcessor added

### PR DESCRIPTION
**Pull request by kuba9392**

### CODE REVIEW

File: src/SAREhub/Client/Processor/LogProcessor.php [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+class LogProcessor implements Processor,LoggerAwareInterface
+{
+    /**
+     * @var LoggerInterface
+     */
+    private $logger;
+
+    public function process(Exchange $exchange)
+    {
+        $this->logger->info($exchange);
```
Comment (Mararok) [inline src/SAREhub/Client/Processor/LogProcessor.php]: Add setter and getter for log level and use generic log method
$this->logger->log($level, $message, array $context = ["exchange" => $exchange]);
Use id(IdAware) for message parameter

File: src/SAREhub/Client/Processor/LogProcessor.php [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+use Psr\Log\LoggerAwareInterface;
+use Psr\Log\LoggerInterface;
+use SAREhub\Client\Message\Exchange;
+
+class LogProcessor implements Processor,LoggerAwareInterface
+{
+    /**
+     * @var LoggerInterface
+     */
+    private $logger;
```
Comment (Mararok) [inline src/SAREhub/Client/Processor/LogProcessor.php:18]: Logger instance from constructor

File: src/SAREhub/Client/Processor/LogProcessor.php [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+
+namespace SAREhub\Client\Processor;
+
+
+use Psr\Log\LoggerAwareInterface;
+use Psr\Log\LoggerInterface;
+use SAREhub\Client\Message\Exchange;
+
+class LogProcessor implements Processor,LoggerAwareInterface
```
Comment (Mararok) [inline src/SAREhub/Client/Processor/LogProcessor.php]: Implement IdAware interface

**Review by Mararok — CHANGES_REQUESTED** [review]

File: src/SAREhub/Client/Processor/LogProcessor.php (hunk 2 above)
Comment (kuba9392) [reply src/SAREhub/Client/Processor/LogProcessor.php:18]: ready

File: src/SAREhub/Client/Processor/LogProcessor.php [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+    public function setLogLevel(string $logLevel): self
+    {
+        $this->logLevel = $logLevel;
+        return $this;
+    }
+
+    public function process(Exchange $exchange)
+    {
+        $this->logger->log($this->logLevel, "logProcessor output[id: " . ($this->getId() ?? 'null') . "]", [
```
Comment (Mararok) [inline src/SAREhub/Client/Processor/LogProcessor.php]: $this->logger->log($this->logLevel, (string)$this, [ "exchange" => $exchange ]);

File: src/SAREhub/Client/Processor/LogProcessor.php [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    private $id;
+
+    private $logLevel = "debug";
+
+    public function __construct(LoggerInterface $logger)
+    {
+        $this->logger = $logger;
+    }
+
+    public static function newInstance(LoggerInterface $logger)
```
Comment (Mararok) [inline src/SAREhub/Client/Processor/LogProcessor.php]: remove

File: src/SAREhub/Client/Processor/LogProcessor.php [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+class LogProcessor implements Processor, LoggerAwareInterface, IdAware
+{
+    /**
+     * @var LoggerInterface
+     */
+    private $logger;
+
+    private $id;
+
+    private $logLevel = "debug";
```
Comment (Mararok) [inline src/SAREhub/Client/Processor/LogProcessor.php:22]: add getter

**Review by Mararok — CHANGES_REQUESTED** [review]

File: src/SAREhub/Client/Processor/LogProcessor.php (hunk 6 above)
Comment (kuba9392) [reply src/SAREhub/Client/Processor/LogProcessor.php:22]: ready

File: src/SAREhub/Client/Processor/LogProcessor.php (hunk 5 above)
Comment (kuba9392) [reply src/SAREhub/Client/Processor/LogProcessor.php]: ready

File: src/SAREhub/Client/Processor/LogProcessor.php (hunk 4 above)
Comment (kuba9392) [reply src/SAREhub/Client/Processor/LogProcessor.php]: ready

File: tests/unit/SAREhub/Client/Processor/LogProcessorTest.php [hunk 7]
```diff
@@ -40,7 +40,7 @@ public function testLogExchangeWhenLogLevelIsFoundAndIdIsNull()
     {
         $logLevel = "error";
         $this->processor->setLogLevel($logLevel);
-        $this->logger->allows("log")->with($logLevel, "logProcessor output[id: null]", ["exchange" => $this->exchange])->once();
+        $this->logger->allows("log")->with($logLevel, (string)$this->processor, ["exchange" => $this->exchange])->once();
```
Comment (Mararok) [inline tests/unit/SAREhub/Client/Processor/LogProcessorTest.php]: $this->logger->expects("log")...

File: src/SAREhub/Client/Processor/LogProcessor.php [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+
+    public function __construct(LoggerInterface $logger)
+    {
+        $this->logger = $logger;
+    }
+
+    public function setLogLevel(string $logLevel)
+    {
+        $this->logLevel = $logLevel;
```
Comment (Mararok) [inline src/SAREhub/Client/Processor/LogProcessor.php]: check level is valid with LogLevel class constants

**Review by Mararok — CHANGES_REQUESTED** [review]

File: src/SAREhub/Client/Processor/LogProcessor.php (hunk 8 above)
Comment (kuba9392) [reply src/SAREhub/Client/Processor/LogProcessor.php]: ready

File: tests/unit/SAREhub/Client/Processor/LogProcessorTest.php (hunk 7 above)
Comment (kuba9392) [reply tests/unit/SAREhub/Client/Processor/LogProcessorTest.php]: ready

File: src/SAREhub/Client/Processor/LogProcessor.php [hunk 9]
```diff
@@ -55,8 +55,18 @@ public function getLogLevel(): string
 
     public function setLogLevel(string $logLevel)
     {
+        $this->isValidLogLevel($logLevel);
         $this->logLevel = $logLevel;
-        return $this;
+    }
+
+    private function isValidLogLevel(string $logLevel): void
```
Comment (Mararok) [inline src/SAREhub/Client/Processor/LogProcessor.php]: too complicated 

if (!in_array(strtolower($logLevel), $logLevelReflection->getConstants())) {
  // throw exception
}

Comment (Mararok) [reply src/SAREhub/Client/Processor/LogProcessor.php]: add test for logLevel valid value but is like "Info" or "iNFO"

**Review by Mararok — CHANGES_REQUESTED** [review]

File: tests/unit/SAREhub/Client/Processor/LogProcessorTest.php [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+
+    public function testLogExchangeWhenLogLevelIsFoundAndIdIsNull()
+    {
+        $logLevel = "error";
+        $this->processor->setLogLevel($logLevel);
+        $this->logger->expects("log")->with($logLevel, (string)$this->processor, ["exchange" => $this->exchange])->once();
+        $this->processor->process($this->exchange);
+    }
+
+    public function testLogExchangeWhenLogLevelIsFoundAndIdIsSet()
```
Comment (Mararok) [inline tests/unit/SAREhub/Client/Processor/LogProcessorTest.php]: testProcessWhenIdIsSetThenLog

File: tests/unit/SAREhub/Client/Processor/LogProcessorTest.php [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+    private $exchange;
+
+    public function setUp()
+    {
+        $this->logger = \Mockery::mock(LoggerInterface::class);
+        $this->processor = new LogProcessor($this->logger);
+        $this->exchange = $this->createExchange();
+    }
+
+    public function testLogExchangeWhenLogLevelIsFoundAndIdIsNull()
```
Comment (Mararok) [inline tests/unit/SAREhub/Client/Processor/LogProcessorTest.php]: testProcessWhenIdIsEmptyThenLog

**Review by Mararok — CHANGES_REQUESTED** [review]

**Review by Mararok — APPROVED** [review]

**Pull request merged**